copy/copy.go

```diff
@@ -231,16 +232,29 @@ func Image(ctx context.Context, policyContext *signature.PolicyContext, destRef,
 		progressOutput:   progressOutput,
 		progressInterval: options.ProgressInterval,
 		progress:         options.Progress,
-		copyInParallel:   copyInParallel,
 		// FIXME? The cache is used for sources and destinations equally, but we only have a SourceCtx and DestinationCtx.
 		// For now, use DestinationCtx (because blob reuse changes the behavior of the destination side more); eventually
 		// we might want to add a separate CommonCtx — or would that be too confusing?
 		blobInfoCache:         internalblobinfocache.FromBlobInfoCache(blobinfocache.DefaultCache(options.DestinationCtx)),
 		ociDecryptConfig:      options.OciDecryptConfig,
 		ociEncryptConfig:      options.OciEncryptConfig,
-		maxParallelDownloads:  options.MaxParallelDownloads,
 		downloadForeignLayers: options.DownloadForeignLayers,
+		layerCopySemaphore:    options.GlobalParallelDownloadsSemaphore,
+	}
+
+	// Set the layerCopySemaphore if we can copy layers in parallel.
+	if dest.HasThreadSafePutBlob() && rawSource.HasThreadSafeGetBlob() {
+		if c.layerCopySemaphore == nil {
+			max := options.MaxParallelDownloads
+			if max == 0 {
+				max = maxParallelDownloads
+			}
+			c.layerCopySemaphore = semaphore.NewWeighted(int64(max))
+		}
+	} else {
+		c.layerCopySemaphore = semaphore.NewWeighted(int64(1))
```

Comment:
With this implementation, with a global semaphore for 5 copies, initiating 20 copies will cause only 5 layer copies if in parallel, but 20 if not parallel. I think this should acquire the global semaphore here, and create a local 1-weight semaphore for local synchronization.

OTOH I can see the argument for “you started 20 copies, you get 20 copies”…

Reply:
Fair point. Guess we should do what you suggest :+1: